cmd/snap-update-ns/change.go

```diff
@@ -591,106 +586,124 @@ func neededChangesOld(currentProfile, desiredProfile *osutil.MountProfile) []*Ch
+	// 1. Perform the mounts for the "overname" entries
+	// 2. Perform the mounts for the entries which need a mimic
+	// 3. Perform all the remaining desired mounts
 
-	// Clean the directory part of both profiles. This is done so that we can
-	// easily test if a given directory is a subdirectory with
-	// strings.HasPrefix coupled with an extra slash character.
-	for i := range current {
-		current[i].Dir = filepath.Clean(current[i].Dir)
-	}
-	for i := range desired {
-		desired[i].Dir = filepath.Clean(desired[i].Dir)
+	var newDesiredEntries []osutil.MountEntry
+	// Indexed by mount point path.
+	addedDesiredEntries := make(map[string]bool)
+	ensureDesiredEntry := func(entry osutil.MountEntry) {
+		if !addedDesiredEntries[entry.Dir] {
+			newDesiredEntries = append(newDesiredEntries, entry)
+			addedDesiredEntries[entry.Dir] = true
+		}
 	}
 
-	// Sort both lists by directory name with implicit trailing slash.
-	sort.Sort(byOriginAndMagicDir(current))
-	sort.Sort(byOriginAndMagicDir(desired))
+	// Create a map of the target directories (mimics) needed for the  visited
+	// entries
+	affectedTargetCreationDirs := map[string][]osutil.MountEntry{}
+	for _, entry := range desiredNotReused {
+		if entry.XSnapdOrigin() == "overname" {
+			ensureDesiredEntry(entry)
+		}
 
-	// We are now ready to compute the necessary mount changes.
-	var changes []*Change
+		parentTargetDir := filepath.Dir(entry.Dir)
+		affectedTargetCreationDirs[parentTargetDir] = append(affectedTargetCreationDirs[parentTargetDir], entry)
+	}
 
-	// Unmount entries in reverse order, so that the most nested element is
-	// always processed first.
-	for i := len(current) - 1; i >= 0; i-- {
-		var entry osutil.MountEntry = current[i]
-		entry.Options = append([]string(nil), entry.Options...)
-		switch {
-		case entry.XSnapdSynthetic() && entry.Type == "tmpfs":
-			// Synthetic changes are rooted under a tmpfs, detach that tmpfs to
-			// remove them all.
-			if !entry.XSnapdDetach() {
-				entry.Options = append(entry.Options, osutil.XSnapdDetach())
-			}
-		case entry.XSnapdSynthetic():
-			// Consume all other syn ethic entries without emitting either a
-			// mount, unmount or keep change.  This relies on the fact that all
-			// synthetic mounts are created by a mimic underneath a tmpfs that
-			// is detached, as coded above.
-			continue
-		case entry.OptBool("rbind") || entry.Type == "tmpfs":
-			// Recursive bind mounts and non-mimic tmpfs mounts need to be
-			// detached because they can contain other mount points that can
-			// otherwise propagate in a self-conflicting way.
-			if !entry.XSnapdDetach() {
-				entry.Options = append(entry.Options, osutil.XSnapdDetach())
+	if len(affectedTargetCreationDirs) != 0 {
+		entriesForMimicDir := map[string][]osutil.MountEntry{}
+		for parentTargetDir, entriesNeedingDir := range affectedTargetCreationDirs {
+			// first check if any of the mount entries for the changes will require
+			// creating a mimic
```

Comment:
whether the target is read-only or not plays a role as well? I suppose the sorting doesn't hurt but I think the comment should be clearer that these are potential mimics and not the exact set? or am I missing something

Reply:
I don't think it does; AFAIK a bind mound does not alter the permissions of the mounted entry, and retains them regardless of the permissions of the mount point. Though I'm not sure if this answers your question in any way :-)

Reply:
Yes Alberto is correct, if the target is read-only we can still mount on top of it just fine, it's really only if the target itself doesn't exist _and_ we can't create it because it's read-only that this matters. 

I do see a potential issue here though, in NeededChanges right now in this branch we are not checking if the target file can be created (in the case it doesn't exist but it is writable). The previous code would dynamically just create the file/directory/symlink later in change.Perform() (called after NeededChanges) and if that failed create the mimic late, but here we are assuming that the mimic will always be created if the target does not exist which is not precisely true. I suppose what we should probably do is actually try to create the file here and then if that succeeds we can skip consideration of this mount entry because it won't require a writable mimic. 

I think the impact of not doing that is somewhat minimal though it is confusing enough we should probably fix it lest this code get even more complicated with less knowledge about what is going on

Reply:
I wrote code to do this but my first stab at it is breaking parallel installs sharing content interface files so I need to work on it a bit more, hopefully tomorrow I will get something working

Reply:
I am a bit confused as I'm not sure what you mean about trying to create the file? Doing final/full work that Perform should do here seems strange?

Anyway, my comment was mostly about the code comment itself, about potential mimic vs actual mimic (your *which is not precisely true*). As I wrote sorting shouldn't have a bad effect either way though.

Reply:
Okay after various attempts I was not able to come up with a set of potential mimics that end up not being real mimics that would result in an incorrect ordering of actual mimic creation that we can't undo (like the original bug robust namespace updates was about).

Reply:
Also updated the doc-comment here